File: sdk/cli/src/manifest.ts
/** Loading and validating manifest.json, with errors that name the offending field. */

import { readFileSync } from "node:fs";
import { CliError } from "./log.js";

export const CAPABILITIES = [
  "system",
  "power",
  "display",
  "sensors",
  "notifications",
  "apps",
  "wallet",
] as const;

export type Capability = (typeof CAPABILITIES)[number];

export interface Manifest {
  id: string;
  name: string;
  version: string;
  sdk: string;
  type: "app" | "watchface";
  entry: string;
  icon?: string;
  capabilities: Capability[];
  author?: string;
  description?: string;
}

const ID_PATTERN = /^[a-z][a-z0-9]*(\.[a-z0-9][a-z0-9-]*)+$/;
const VERSION_PATTERN = /^\d+\.\d+\.\d+(?:[-+][0-9A-Za-z.-]+)?$/;

export function parseManifest(text: string, source: string): Manifest {
  let value: unknown;
  try {
    value = JSON.parse(text);
  } catch (error) {
    throw new CliError(`${source} is not valid JSON: ${(error as Error).message}`, {
      hint: "Open the file and look for a trailing comma or an unquoted key.",
    });
  }
  return validateManifest(value, source);
}

export function validateManifest(value: unknown, source: string): Manifest {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new CliError(`${source} must contain a JSON object.`);
  }
  const raw = value as Record<string, unknown>;
  const problems: string[] = [];

  const requireString = (field: string): string => {
    const found = raw[field];
    if (typeof found !== "string" || found.trim() === "") {
      problems.push(`"${field}" is required and must be a non-empty string`);
      return "";
    }
    return found;
  };

  const id = requireString("id");
  if (id && !ID_PATTERN.test(id)) {
    problems.push(`"id" must be reverse-DNS, for example tech.solwear.watchface (got "${id}")`);
  }

  const name = requireString("name");
  const version = requireString("version");
  if (version && !VERSION_PATTERN.test(version)) {
    problems.push(`"version" must be semantic versioning, for example 1.0.0 (got "${version}")`);
  }

  const sdk = typeof raw["sdk"] === "string" ? (raw["sdk"] as string) : "";
  if (!sdk) problems.push('"sdk" is required, for example "0.1"');

  const type = raw["type"];
  if (type !== "app" && type !== "watchface") {
    problems.push(`"type" must be either "app" or "watchface" (got ${JSON.stringify(type)})`);
  }

  const entry = typeof raw["entry"] === "string" && raw["entry"] ? (raw["entry"] as string) : "index.html";
  if (entry.startsWith("/") || entry.includes("..")) {
    problems.push(`"entry" must be a relative path inside the package (got "${entry}")`);
  }

  const capabilitiesRaw = raw["capabilities"];
  const capabilities: Capability[] = [];
  if (!Array.isArray(capabilitiesRaw)) {
    problems.push('"capabilities" is required and must be an array, use [] if the app needs nothing');
  } else {
    for (const capability of capabilitiesRaw) {
      if (typeof capability !== "string" || !(CAPABILITIES as readonly string[]).includes(capability)) {
        problems.push(
          `"${String(capability)}" is not a capability. Valid values: ${CAPABILITIES.join(", ")}`,
        );
      } else if (capabilities.includes(capability as Capability)) {
        problems.push(`"${capability}" is listed twice in "capabilities"`);
      } else {
        capabilities.push(capability as Capability);
      }
    }
  }

  const icon = typeof raw["icon"] === "string" ? (raw["icon"] as string) : undefined;
  if (icon && (icon.startsWith("/") || icon.includes(".."))) {
    problems.push(`"icon" must be a relative path inside the package (got "${icon}")`);
  }

  if (problems.length > 0) {
    throw new CliError(
      `${source} is not a valid SolWear manifest:\n` + problems.map((p) => `    - ${p}`).join("\n"),
      { hint: "Section 5 of docs/ARCHITECTURE.md describes every field." },
    );
  }

  const manifest: Manifest = {
    id,
    name,
    version,
    sdk,
    type: type as "app" | "watchface",
    entry,
    capabilities,
  };
  if (icon) manifest.icon = icon;
  if (typeof raw["author"] === "string") manifest.author = raw["author"] as string;
  if (typeof raw["description"] === "string") manifest.description = raw["description"] as string;
  return manifest;
}

export function readManifest(path: string): Manifest {
  let text: string;
  try {
    text = readFileSync(path, "utf8");
  } catch {
    throw new CliError(`No manifest.json at ${path}.`, {
      hint: "Run this command from an app directory, or create one with `solwear new <name>`.",
    });
  }
  return parseManifest(text, path);
}

/** The file name a package gets: `<id>-<version>.swa`. */
export function packageFileName(manifest: Manifest): string {
  return `${manifest.id}-${manifest.version}.swa`;
}

File: os/solweard/src/manifest.rs
//! `manifest.json` parsing and validation for `.swa` packages.

use serde::{Deserialize, Serialize};

/// Capability names recognised by the daemon. They map one to one onto the
/// method namespace prefix used in the JSON-RPC API.
pub const CAPABILITIES: &[&str] = &[
    "system",
    "power",
    "display",
    "sensors",
    "notifications",
    "apps",
    "wallet",
];

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct Manifest {
    pub id: String,
    pub name: String,
    pub version: String,
    pub sdk: String,
    #[serde(rename = "type")]
    pub app_type: String,
    pub entry: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub icon: Option<String>,
    #[serde(default)]
    pub capabilities: Vec<String>,
    #[serde(default)]
    pub author: String,
    #[serde(default)]
    pub description: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ManifestError(pub String);

impl std::fmt::Display for ManifestError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl std::error::Error for ManifestError {}

fn err(message: impl Into<String>) -> ManifestError {
    ManifestError(message.into())
}

impl Manifest {
    pub fn parse(bytes: &[u8]) -> Result<Manifest, ManifestError> {
        let manifest: Manifest = serde_json::from_slice(bytes)
            .map_err(|e| err(format!("manifest.json is not valid: {e}")))?;
        manifest.validate()?;
        Ok(manifest)
    }

    pub fn validate(&self) -> Result<(), ManifestError> {
        validate_app_id(&self.id)?;
        if self.name.trim().is_empty() {
            return Err(err("manifest `name` must not be empty"));
        }
        validate_version(&self.version)?;
        if self.sdk.trim().is_empty() {
            return Err(err("manifest `sdk` must not be empty"));
        }
        if !matches!(self.app_type.as_str(), "app" | "watchface") {
            return Err(err(format!(
                "manifest `type` must be `app` or `watchface`, got `{}`",
                self.app_type
            )));
        }
        validate_relative_path(&self.entry, "entry")?;
        if let Some(icon) = &self.icon {
            validate_relative_path(icon, "icon")?;
        }
        for capability in &self.capabilities {
            if !CAPABILITIES.contains(&capability.as_str()) {
                return Err(err(format!("unknown capability `{capability}`")));
            }
        }
        Ok(())
    }

    pub fn has_capability(&self, capability: &str) -> bool {
        self.capabilities.iter().any(|c| c == capability)
    }
}

/// App ids are reverse-DNS, lowercase, and immutable. Keeping the character set
/// tight means an id is always safe to use as a directory name and as a URL
/// path segment.
pub fn validate_app_id(id: &str) -> Result<(), ManifestError> {
    if id.is_empty() || id.len() > 128 {
        return Err(err("app id must be between 1 and 128 characters"));
    }
    if id == "tech.solwear.shell" {
        return Err(err("app id `tech.solwear.shell` is reserved by the system"));
    }
    let labels: Vec<&str> = id.split('.').collect();
    if labels.len() < 2 {
        return Err(err(format!(
            "app id `{id}` must be reverse-DNS, e.g. tech.solwear.app"
        )));
    }
    for label in labels {
        if label.is_empty() {
            return Err(err(format!("app id `{id}` has an empty label")));
        }
        if !label.chars().next().is_some_and(|c| c.is_ascii_lowercase()) {
            return Err(err(format!(
                "app id `{id}` has a label that does not start with a lowercase letter"
            )));
        }
        if !label
            .chars()
            .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '-')
        {
            return Err(err(format!(
                "app id `{id}` may only contain lowercase letters, digits, hyphens and dots"
            )));
        }
    }
    Ok(())
}

/// Versions are `major.minor.patch` with an optional pre-release suffix.
fn validate_version(version: &str) -> Result<(), ManifestError> {
    let core = version.split(['-', '+']).next().unwrap_or("");
    let parts: Vec<&str> = core.split('.').collect();
    if parts.len() != 3
        || parts
            .iter()
            .any(|p| p.is_empty() || !p.chars().all(|c| c.is_ascii_digit()))
    {
        return Err(err(format!(
            "manifest `version` must be semver, got `{version}`"
        )));
    }
    Ok(())
}

/// Paths inside a package must stay inside the package.
pub fn validate_relative_path(path: &str, field: &str) -> Result<(), ManifestError> {
    if path.trim().is_empty() {
        return Err(err(format!("manifest `{field}` must not be empty")));
    }
    if path.starts_with('/') || path.starts_with('\\') || path.contains(':') {
        return Err(err(format!(
            "manifest `{field}` must be a relative path, got `{path}`"
        )));
    }
    if path
        .split(['/', '\\'])
        .any(|segment| segment == ".." || segment == ".")
    {
        return Err(err(format!(
            "manifest `{field}` must not traverse directories: `{path}`"
        )));
    }
    Ok(())
}
